**Sunita Tebo: Project 2- Preparing the data for machine learning**

In [230]:
#import the necessary libraries
import pandas as pd
import numpy as np

In [231]:
#load the dataset
df1=pd.read_csv("historical_stocks.csv")
df2= pd.read_csv("historical_stock_prices.csv")

In [232]:
#verify the structure of the dataset
print(df1.info())
print(df2.info())

print(df1.shape)
print(df2.shape)

<class 'pandas.DataFrame'>
RangeIndex: 6460 entries, 0 to 6459
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype
---  ------    --------------  -----
 0   ticker    6460 non-null   str  
 1   exchange  6460 non-null   str  
 2   name      6460 non-null   str  
 3   sector    5020 non-null   str  
 4   industry  5020 non-null   str  
dtypes: str(5)
memory usage: 252.5 KB
None
<class 'pandas.DataFrame'>
RangeIndex: 20973889 entries, 0 to 20973888
Data columns (total 8 columns):
 #   Column     Dtype  
---  ------     -----  
 0   ticker     str    
 1   open       float64
 2   close      float64
 3   adj_close  float64
 4   low        float64
 5   high       float64
 6   volume     int64  
 7   date       str    
dtypes: float64(5), int64(1), str(2)
memory usage: 1.3 GB
None
(6460, 5)
(20973889, 8)


In [233]:
#converting date datatype from string to datetime
df2['date'] = pd.to_datetime(df2['date'])

In [234]:
#verify the changes
df2.info()

<class 'pandas.DataFrame'>
RangeIndex: 20973889 entries, 0 to 20973888
Data columns (total 8 columns):
 #   Column     Dtype         
---  ------     -----         
 0   ticker     str           
 1   open       float64       
 2   close      float64       
 3   adj_close  float64       
 4   low        float64       
 5   high       float64       
 6   volume     int64         
 7   date       datetime64[us]
dtypes: datetime64[us](1), float64(5), int64(1), str(1)
memory usage: 1.3 GB


In [103]:
#sort the stock prices dataset(df2) by tick and date.
df2 = df2.sort_values(["ticker", "date"])

In [235]:
#verify the result
print(df2.head(20))

   ticker  open  close  adj_close   low  high   volume       date
0     AHH 11.50  11.58       8.49 11.25 11.68  4633900 2013-05-08
1     AHH 11.66  11.55       8.47 11.50 11.66   275800 2013-05-09
2     AHH 11.55  11.60       8.51 11.50 11.60   277100 2013-05-10
3     AHH 11.63  11.65       8.54 11.55 11.65   147400 2013-05-13
4     AHH 11.60  11.53       8.46 11.50 11.60   184100 2013-05-14
5     AHH 11.60  11.60       8.51 11.54 11.60    76800 2013-05-15
6     AHH 11.62  11.74       8.61 11.54 11.74   170300 2013-05-16
7     AHH 11.70  11.76       8.63 11.70 11.85   305400 2013-05-17
8     AHH 11.76  11.73       8.60 11.63 11.83    46800 2013-05-20
9     AHH 11.76  11.83       8.68 11.61 11.84    77000 2013-05-21
10    AHH 11.84  11.75       8.62 11.68 11.84    90200 2013-05-22
11    AHH 11.64  11.61       8.52 11.54 11.79    75400 2013-05-23
12    AHH 11.57  11.70       8.58 11.55 11.73    63400 2013-05-24
13    AHH 11.73  11.58       8.49 11.57 11.78   212400 2013-05-28
14    AHH 

In [236]:
#verify duplicated values for the ticker and date
df2.duplicated(subset=['ticker', 'date']).sum()

np.int64(0)

No duplicate in the ticker and dates

In [237]:
#check which tickers don't have prices
df1.loc[df1['ticker'].isin(df2['ticker']),'ticker'].head(10)

0       PIH
1     PIHPP
2      TURN
3      FLWS
4      FCCY
5      SRCE
6      VNET
7      TWOU
8      JOBS
10     AVHI
Name: ticker, dtype: str

In [238]:
#in project 1 we verified a false header in df1, called symbol
df1[df1['ticker']=='Symbol']

,ticker,exchange,name,sector,industry
74,Symbol,NYSE,NAME,SECTOR,INDUSTRY


In [239]:
#verify if there are other sector with symbol
df1['sector'].value_counts(dropna=False)

sector
NaN                      1440
FINANCE                  1022
CONSUMER SERVICES         796
HEALTH CARE               784
TECHNOLOGY                607
CAPITAL GOODS             352
ENERGY                    286
PUBLIC UTILITIES          273
BASIC INDUSTRIES          272
CONSUMER NON-DURABLES     224
CONSUMER DURABLES         144
MISCELLANEOUS             139
TRANSPORTATION            120
SECTOR                      1
Name: count, dtype: int64

Based on the output, there is just one sector with symbol and therefor the next step will be to verify if df2 has a ticker called symbol


In [240]:
nan_sector = df1[df1['sector'].isna()]
nan_sector[['ticker', 'exchange', 'name', 'industry']].isna().sum()

ticker         0
exchange       0
name           0
industry    1440
dtype: int64

In [241]:
nan_sector[['ticker', 'exchange', 'name', 'industry']].head(20)

,ticker,exchange,name,industry
19,ABP,NASDAQ,ABPRO CORPORATION,NaN
42,SQZZ,NASDAQ,ACTIVE ALTS CONTRARIAN ETF,NaN
62,ACT,NASDAQ,ADVISORSHARES VICE ETF,NaN
100,ABDC,NASDAQ,ALCENTRA CAPITAL CORP.,NaN
124,SMCP,NASDAQ,ALPHAMARK ACTIVELY MANAGED SMALL CAP ETF,NaN
127,SWIN,NASDAQ,ALPS/DORSEY WRIGHT SECTOR MOMENTUM ETF,NaN
151,ACSF,NASDAQ,"AMERICAN CAPITAL SENIOR FLOATING, LTD.",NaN
172,IBUY,NASDAQ,AMPLIFY ONLINE RETAIL ETF,NaN
191,AINV,NASDAQ,APOLLO INVESTMENT CORPORATION,NaN
208,AQB,NASDAQ,"AQUABOUNTY TECHNOLOGIES, INC.",NaN


Since the NaN industry has other features, such as exchange, ticker and name, i will not delete them for now. however i might replace them with unknown if the ml throws an error.

In [242]:
#verify if there is an actual ticker in df2 called symbol
print(df2[df2["ticker"] == "Symbol"])

Empty DataFrame
Columns: [ticker, open, close, adj_close, low, high, volume, date]
Index: []


Based on this output, the dataframe is empty meaning there is no ticker in the historic price dataset (df2) with ticker named symbol. Therefore, i will be dropping the row for the following reason:
Symbol has no price and no matching data in the price dataset. Therefore, it has no impact on the dataset

In [243]:
#dropping the symbol from df1
df1 = df1[df1["ticker"] != "Symbol"]

In [244]:
#verify if it has been dropped
print("Symbol" in df1["ticker"].values)

False


In [245]:
df1['sector'].value_counts(dropna=False)

sector
NaN                      1440
FINANCE                  1022
CONSUMER SERVICES         796
HEALTH CARE               784
TECHNOLOGY                607
CAPITAL GOODS             352
ENERGY                    286
PUBLIC UTILITIES          273
BASIC INDUSTRIES          272
CONSUMER NON-DURABLES     224
CONSUMER DURABLES         144
MISCELLANEOUS             139
TRANSPORTATION            120
Name: count, dtype: int64

the row has been deleted

***unusal observationform project 1***


In [246]:
##use the describe function to find the mean, median, count, maximum and minimum
pd.set_option('display.float_format', '{:,.2f}'.format)

df2[['open', 'high', 'low', 'close', 'volume']].describe()

,open,high,low,close,volume
count,"20,973,889.00","20,973,889.00","20,973,889.00","20,973,889.00","20,973,889.00"
mean,76.06,78.04,74.22,76.11,"1,227,043.03"
std,"2,849.64","2,997.94","2,746.06","2,870.16","13,166,864.47"
min,0.00,0.00,0.00,0.00,1.00
25%,7.50,7.63,7.36,7.50,"22,100.00"
50%,15.45,15.66,15.24,15.45,"126,000.00"
75%,29.72,30.10,29.28,29.72,"607,400.00"
max,"2,034,000.00","2,070,000.00","1,440,000.00","1,779,750.00","4,483,504,400.00"


mean closing price is $76.11.
median closing price is $15.45.
the difference between the mean and median are $76.11 − $15.45 = $60.66. that is mean is $60.66 higher than the median. this large difference suggests that the closing-price distribution is strongly right-skewed
the variable with the largest standard deviation is volume with Volume = 13,166,864.47

In [247]:
#find the unusally features
highest_close = df2.nlargest(10, 'close')

print(highest_close[['ticker', 'date', 'open', 'high', 'low', 'close','volume']])

         ticker       date         open         high          low  \
12146443   SCON 2000-02-28 1,395,000.00 1,797,750.00 1,368,000.00   
12146459   SCON 2000-02-29 2,034,000.00 2,070,000.00 1,440,000.00   
12146546   SCON 2000-03-13   900,000.00 1,404,000.00   861,750.00   
17889354   CODA 2001-02-05 1,354,500.00 1,736,700.00 1,354,500.00   
2444619    TVIX 2013-02-26 1,375,000.00 1,542,500.00 1,300,000.00   
12146480   SCON 2000-03-02 1,120,500.00 1,368,000.00 1,098,280.75   
12146437   SCON 2000-02-25   873,561.62 1,369,125.00   873,000.00   
17107201   TOPS 2016-08-01   783,000.00 1,276,200.00   783,000.00   
12146465   SCON 2000-03-01 1,368,000.00 1,512,000.00 1,107,000.00   
12146562   SCON 2000-03-14 1,352,250.00 1,354,500.00 1,080,000.00   

                close  volume  
12146443 1,779,750.00     400  
12146459 1,595,250.00     400  
12146546 1,363,500.00     300  
17889354 1,354,500.00      34  
2444619  1,347,500.00     100  
12146480 1,261,125.00     200  
12146437 1,215,0

The largest adjusted closing values in your results are all for AAN with an adjusted closing price of about 18.95 qunatile while the actual closing price is about 1.30. There are huge difference in comparing the close price and the adjusted close price for the 10 largest prices. close prices are around $1.27–$1.30, but the adj_close values are approximately $18,949,618,910,812,401,664

The unusual values occur in 1987. Again, this could be an error or a change in circumstances not necessarily an error. for now i would not delete them or try to standardize them

In [248]:
#Find the largest adjusted closing values
highest_adj_close = df2.nlargest(10, 'adj_close')

print(highest_adj_close[
    ['ticker', 'date', 'close', 'adj_close']
])

        ticker       date  close                     adj_close
6174033    AAN 1987-02-27   1.30 18,949,618,910,812,401,664.00
6174039    AAN 1987-03-10   1.30 18,949,618,910,812,401,664.00
6174040    AAN 1987-03-11   1.30 18,949,618,910,812,401,664.00
6174041    AAN 1987-03-12   1.30 18,949,618,910,812,401,664.00
6174044    AAN 1987-03-17   1.30 18,949,618,910,812,401,664.00
6174045    AAN 1987-03-18   1.30 18,949,618,910,812,401,664.00
6174046    AAN 1987-03-19   1.30 18,949,618,910,812,401,664.00
6174049    AAN 1987-03-25   1.30 18,949,618,910,812,401,664.00
6174051    AAN 1987-03-27   1.29 18,814,275,626,503,000,064.00
6174050    AAN 1987-03-26   1.27 18,543,558,271,558,500,352.00


Note that the adjusted closing price does not only provide a reflection of what the close price is, but it also modifies the close price value to account for corporate actions like dividends and stock splits, which provide a more accurate reflection of historical performance. 

Therefore, deciding on what to do with this unusually high adjusted closed price is crucial. 

If i Retain it, that is Keep the values without identifying them. it will distored the analyses as teh model might be unfair with the huge numbers thinking they are more important.

If i remove them, i will be loosing a lot of data becasue from project one, i verified to see if these could b=have been error that orrured only within a particular dates but it was not the case as it was dertermine that the unusual values were through out the dataset and no specific period of high adj_ closing price. see the output of the next function to confirm

In [249]:
#are the unsul values concentrated around certain periods?
print(
    highest_adj_close[
        ['ticker', 'date', 'close', 'adj_close']
    ].sort_values('date')
)

        ticker       date  close                     adj_close
6174033    AAN 1987-02-27   1.30 18,949,618,910,812,401,664.00
6174039    AAN 1987-03-10   1.30 18,949,618,910,812,401,664.00
6174040    AAN 1987-03-11   1.30 18,949,618,910,812,401,664.00
6174041    AAN 1987-03-12   1.30 18,949,618,910,812,401,664.00
6174044    AAN 1987-03-17   1.30 18,949,618,910,812,401,664.00
6174045    AAN 1987-03-18   1.30 18,949,618,910,812,401,664.00
6174046    AAN 1987-03-19   1.30 18,949,618,910,812,401,664.00
6174049    AAN 1987-03-25   1.30 18,949,618,910,812,401,664.00
6174050    AAN 1987-03-26   1.27 18,543,558,271,558,500,352.00
6174051    AAN 1987-03-27   1.29 18,814,275,626,503,000,064.00


In [250]:
print(
    highest_adj_close[
        ['ticker', 'date', 'close', 'adj_close']
    ].sort_values('date', ascending=False)
)

        ticker       date  close                     adj_close
6174051    AAN 1987-03-27   1.29 18,814,275,626,503,000,064.00
6174050    AAN 1987-03-26   1.27 18,543,558,271,558,500,352.00
6174049    AAN 1987-03-25   1.30 18,949,618,910,812,401,664.00
6174046    AAN 1987-03-19   1.30 18,949,618,910,812,401,664.00
6174045    AAN 1987-03-18   1.30 18,949,618,910,812,401,664.00
6174044    AAN 1987-03-17   1.30 18,949,618,910,812,401,664.00
6174041    AAN 1987-03-12   1.30 18,949,618,910,812,401,664.00
6174040    AAN 1987-03-11   1.30 18,949,618,910,812,401,664.00
6174039    AAN 1987-03-10   1.30 18,949,618,910,812,401,664.00
6174033    AAN 1987-02-27   1.30 18,949,618,910,812,401,664.00


based on this, i will flag the high prices.

In [251]:
#Flaging the high adj close price using the IDR method
Q1 = df2['adj_close'].quantile(0.25)
Q3 = df2['adj_close'].quantile(0.75)

IQR = Q3 - Q1

upper_limit = Q3 + 1.5 * IQR

df2['adj_close_flag'] = df2['adj_close'] > upper_limit

In [252]:
print(
    df2[df2['adj_close_flag']][
        ['ticker', 'date', 'close', 'adj_close']
    ]
)

         ticker       date  close  adj_close
28471      VIAV 1999-03-08  55.18      55.18
28515      VIAV 1999-03-10  56.39      56.39
28536      VIAV 1999-03-11  57.17      57.17
28557      VIAV 1999-03-12  56.81      56.81
28579      VIAV 1999-03-15  56.85      56.85
...         ...        ...    ...        ...
20961363   SPEX 2014-03-28  57.00      57.00
20961364   SPEX 2014-03-31  56.81      56.81
20961365   SPEX 2014-04-01  57.19      57.19
20961403   SPEX 2014-05-27  56.43      56.43
20961404   SPEX 2014-05-28  65.74      65.74

[1775722 rows x 4 columns]


this shows that out of the 4 columns, 1775722 rows are flagged

In [253]:
#verify the count of flagged adj close price
print(df2['adj_close_flag'].value_counts())

adj_close_flag
False    19198167
True      1775722
Name: count, dtype: int64


Flagging these high values is important because, no potentially valid stock data is lose and it gives more room for investigation along the process.

I do not assume that any unsual data is an error especailly because there is no clear partten that this huge data difference is from a specific time period.

***Part 2 — Create Useful Features***


In [254]:
#Sort by ticker and date first
df2 = df2.sort_values(['ticker', 'date'])

In [255]:
#Create price_change colume which is change in closing price from the previous trading day
df2['price_change'] = df2.groupby('ticker')['close'].diff()

In [257]:
#show teh output
print(df2[['ticker', 'date', 'close', 'price_change']].head(10))

         ticker       date  close  price_change
13766141      A 1999-11-18  31.47           NaN
13766149      A 1999-11-19  28.88         -2.59
13766150      A 1999-11-22  31.47          2.59
13766151      A 1999-11-23  28.61         -2.86
13766152      A 1999-11-24  29.37          0.76
13766153      A 1999-11-26  29.46          0.09
13766154      A 1999-11-29  30.13          0.67
13766155      A 1999-11-30  30.18          0.04
13766156      A 1999-12-01  30.71          0.54
13766157      A 1999-12-02  31.56          0.85


The NAN for the first  dates is because there is no previous trading day for the first observation of each ticker
Positive number implies a price increase
Negative number implies a price decrease

In [258]:
# Create a daily return
df2['daily_return'] = df2.groupby('ticker')['close'].pct_change() * 100

The stockes are group per ticker to ensures that each is calculated separately for each stock.

In [259]:
print(df2[['ticker', 'date', 'close', 'price_change','daily_return']].head(20))

         ticker       date  close  price_change  daily_return
13766141      A 1999-11-18  31.47           NaN           NaN
13766149      A 1999-11-19  28.88         -2.59         -8.24
13766150      A 1999-11-22  31.47          2.59          8.98
13766151      A 1999-11-23  28.61         -2.86         -9.09
13766152      A 1999-11-24  29.37          0.76          2.66
13766153      A 1999-11-26  29.46          0.09          0.30
13766154      A 1999-11-29  30.13          0.67          2.28
13766155      A 1999-11-30  30.18          0.04          0.15
13766156      A 1999-12-01  30.71          0.54          1.78
13766157      A 1999-12-02  31.56          0.85          2.77
13766158      A 1999-12-03  31.83          0.27          0.85
13766159      A 1999-12-06  32.73          0.89          2.81
13766160      A 1999-12-07  32.37         -0.36         -1.09
13766161      A 1999-12-08  32.32         -0.04         -0.14
13766162      A 1999-12-09  32.77          0.45          1.38
13766163

In [260]:
#close_lag_1
df2['close_lag_1'] = df2.groupby('ticker')['close'].shift(1)

In [261]:
print(df2[['ticker', 'date', 'close', 'price_change','daily_return', 'close_lag_1']].head(20))

         ticker       date  close  price_change  daily_return  close_lag_1
13766141      A 1999-11-18  31.47           NaN           NaN          NaN
13766149      A 1999-11-19  28.88         -2.59         -8.24        31.47
13766150      A 1999-11-22  31.47          2.59          8.98        28.88
13766151      A 1999-11-23  28.61         -2.86         -9.09        31.47
13766152      A 1999-11-24  29.37          0.76          2.66        28.61
13766153      A 1999-11-26  29.46          0.09          0.30        29.37
13766154      A 1999-11-29  30.13          0.67          2.28        29.46
13766155      A 1999-11-30  30.18          0.04          0.15        30.13
13766156      A 1999-12-01  30.71          0.54          1.78        30.18
13766157      A 1999-12-02  31.56          0.85          2.77        30.71
13766158      A 1999-12-03  31.83          0.27          0.85        31.56
13766159      A 1999-12-06  32.73          0.89          2.81        31.83
13766160      A 1999-12-0

In [262]:
#Create the 5-day moving average
df2['ma_5'] = (
    df2.groupby('ticker')['close']
       .rolling(window=5)
       .mean()
       .reset_index(level=0, drop=True)
)

In [263]:
#Create the 20-day moving average
df2['ma_20'] = (
    df2.groupby('ticker')['close']
       .rolling(window=20)
       .mean()
       .reset_index(level=0, drop=True)
)

In [264]:
print(
    df2[['ticker', 'date', 'close', 'ma_5', 'ma_20']].head(30)
)

         ticker       date  close  ma_5  ma_20
13766141      A 1999-11-18  31.47   NaN    NaN
13766149      A 1999-11-19  28.88   NaN    NaN
13766150      A 1999-11-22  31.47   NaN    NaN
13766151      A 1999-11-23  28.61   NaN    NaN
13766152      A 1999-11-24  29.37 29.96    NaN
13766153      A 1999-11-26  29.46 29.56    NaN
13766154      A 1999-11-29  30.13 29.81    NaN
13766155      A 1999-11-30  30.18 29.55    NaN
13766156      A 1999-12-01  30.71 29.97    NaN
13766157      A 1999-12-02  31.56 30.41    NaN
13766158      A 1999-12-03  31.83 30.88    NaN
13766159      A 1999-12-06  32.73 31.40    NaN
13766160      A 1999-12-07  32.37 31.84    NaN
13766161      A 1999-12-08  32.32 32.16    NaN
13766162      A 1999-12-09  32.77 32.40    NaN
13766163      A 1999-12-10  32.01 32.44    NaN
13766164      A 1999-12-13  32.55 32.40    NaN
13766165      A 1999-12-14  30.76 32.08    NaN
13766175      A 1999-12-15  29.82 31.58    NaN
13766176      A 1999-12-16  33.80 31.79  31.14
13766177     

In [265]:
#Create volatility_20 using the dairly return already created,. Note that Volatility measures how much a stock's returns have been fluctuating
df2['volatility_20'] = (
    df2.groupby('ticker')['daily_return']
       .rolling(window=20)
       .std()
       .reset_index(level=0, drop=True)
)

In [266]:
print(
    df2[['ticker', 'date', 'close', 'volatility_20']].head(30)
)

         ticker       date  close  volatility_20
13766141      A 1999-11-18  31.47            NaN
13766149      A 1999-11-19  28.88            NaN
13766150      A 1999-11-22  31.47            NaN
13766151      A 1999-11-23  28.61            NaN
13766152      A 1999-11-24  29.37            NaN
13766153      A 1999-11-26  29.46            NaN
13766154      A 1999-11-29  30.13            NaN
13766155      A 1999-11-30  30.18            NaN
13766156      A 1999-12-01  30.71            NaN
13766157      A 1999-12-02  31.56            NaN
13766158      A 1999-12-03  31.83            NaN
13766159      A 1999-12-06  32.73            NaN
13766160      A 1999-12-07  32.37            NaN
13766161      A 1999-12-08  32.32            NaN
13766162      A 1999-12-09  32.77            NaN
13766163      A 1999-12-10  32.01            NaN
13766164      A 1999-12-13  32.55            NaN
13766165      A 1999-12-14  30.76            NaN
13766175      A 1999-12-15  29.82            NaN
13766176      A 1999

In [267]:
#Additional feature: high_low_range
df2['high_low_range'] = df2['high'] - df2['low']

I selected high_low_range because it measures the amount of price movement within each trading day. A larger range indicates greater intra day price movement, which can provide additional information about stock behaviour and market activity.

A larger range means the stock had more price movement during that trading day.

In [268]:
#pring the result
print(
    df2[
        ['ticker', 'date', 'close', 'daily_return',
         'volatility_20', 'high_low_range']
    ].head(30)
)

         ticker       date  close  daily_return  volatility_20  high_low_range
13766141      A 1999-11-18  31.47           NaN            NaN            7.15
13766149      A 1999-11-19  28.88         -8.24            NaN            2.28
13766150      A 1999-11-22  31.47          8.98            NaN            2.82
13766151      A 1999-11-23  28.61         -9.09            NaN            2.59
13766152      A 1999-11-24  29.37          2.66            NaN            1.39
13766153      A 1999-11-26  29.46          0.30            NaN            0.54
13766154      A 1999-11-29  30.13          2.28            NaN            1.34
13766155      A 1999-11-30  30.18          0.15            NaN            1.43
13766156      A 1999-12-01  30.71          1.78            NaN            1.12
13766157      A 1999-12-02  31.56          2.77            NaN            1.30
13766158      A 1999-12-03  31.83          0.85            NaN            0.98
13766159      A 1999-12-06  32.73          2.81     

***Part 3 — Combine and Organize the Data***

In [269]:
#select the relevant columns from df1
df1_merge = df1[['ticker', 'exchange', 'sector', 'industry']]

In [270]:
#very the sector column
df1_merge['sector'].value_counts(dropna=False)

sector
NaN                      1440
FINANCE                  1022
CONSUMER SERVICES         796
HEALTH CARE               784
TECHNOLOGY                607
CAPITAL GOODS             352
ENERGY                    286
PUBLIC UTILITIES          273
BASIC INDUSTRIES          272
CONSUMER NON-DURABLES     224
CONSUMER DURABLES         144
MISCELLANEOUS             139
TRANSPORTATION            120
Name: count, dtype: int64

In [271]:
#verfy the industry column
df1_merge['industry'].value_counts(dropna=False)

industry
NaN                                                1440
MAJOR PHARMACEUTICALS                               419
MAJOR BANKS                                         352
REAL ESTATE INVESTMENT TRUSTS                       234
BUSINESS SERVICES                                   223
                                                   ... 
CONSUMER: GREETING CARDS                              1
ELECTRONICS DISTRIBUTION                              1
TOOLS/HARDWARE                                        1
TOBACCO                                               1
GENERAL BLDG CONTRACTORS - NONRESIDENTIAL BLDGS       1
Name: count, Length: 136, dtype: int64

In [272]:
#check whether ticker is unique
print(df1_merge['ticker'].duplicated().sum())

0


In [273]:
#since df2 contains many price records for each ticker, use a left merge so that all price records are retained
df2 = df2.merge(
    df1_merge,
    on='ticker',
    how='left'
)

Check the number of rows for before and after. Note that if the ticker was unique, then the number of rows will not change only the number of columns will

In [274]:
df2.head(20)

,ticker,open,close,adj_close,low,high,volume,date,adj_close_flag,price_change,daily_return,close_lag_1,ma_5,ma_20,volatility_20,high_low_range,exchange,sector,industry
0,A,32.55,31.47,27.49,28.61,35.77,62546300,1999-11-18,False,NaN,NaN,NaN,NaN,NaN,NaN,7.15,NYSE,CAPITAL GOODS,BIOTECHNOLOGY: LABORATORY ANALYTICAL INSTRUMENTS
1,A,30.71,28.88,25.23,28.48,30.76,15234100,1999-11-19,False,-2.59,-8.24,31.47,NaN,NaN,NaN,2.28,NYSE,CAPITAL GOODS,BIOTECHNOLOGY: LABORATORY ANALYTICAL INSTRUMENTS
2,A,29.55,31.47,27.49,28.66,31.47,6577800,1999-11-22,False,2.59,8.98,28.88,NaN,NaN,NaN,2.82,NYSE,CAPITAL GOODS,BIOTECHNOLOGY: LABORATORY ANALYTICAL INSTRUMENTS
3,A,30.40,28.61,25.00,28.61,31.21,5975600,1999-11-23,False,-2.86,-9.09,31.47,NaN,NaN,NaN,2.59,NYSE,CAPITAL GOODS,BIOTECHNOLOGY: LABORATORY ANALYTICAL INSTRUMENTS
4,A,28.70,29.37,25.66,28.61,30.00,4843200,1999-11-24,False,0.76,2.66,28.61,29.96,NaN,NaN,1.39,NYSE,CAPITAL GOODS,BIOTECHNOLOGY: LABORATORY ANALYTICAL INSTRUMENTS
5,A,29.24,29.46,25.74,29.15,29.69,1729400,1999-11-26,False,0.09,0.30,29.37,29.56,NaN,NaN,0.54,NYSE,CAPITAL GOODS,BIOTECHNOLOGY: LABORATORY ANALYTICAL INSTRUMENTS
6,A,29.33,30.13,26.32,29.01,30.36,4074700,1999-11-29,False,0.67,2.28,29.46,29.81,NaN,NaN,1.34,NYSE,CAPITAL GOODS,BIOTECHNOLOGY: LABORATORY ANALYTICAL INSTRUMENTS
7,A,30.04,30.18,26.36,29.28,30.71,4310000,1999-11-30,False,0.04,0.15,30.13,29.55,NaN,NaN,1.43,NYSE,CAPITAL GOODS,BIOTECHNOLOGY: LABORATORY ANALYTICAL INSTRUMENTS
8,A,30.18,30.71,26.83,29.95,31.07,2957300,1999-12-01,False,0.54,1.78,30.18,29.97,NaN,NaN,1.12,NYSE,CAPITAL GOODS,BIOTECHNOLOGY: LABORATORY ANALYTICAL INSTRUMENTS
9,A,31.29,31.56,27.57,30.89,32.19,3069800,1999-12-02,False,0.85,2.77,30.71,30.41,NaN,NaN,1.30,NYSE,CAPITAL GOODS,BIOTECHNOLOGY: LABORATORY ANALYTICAL INSTRUMENTS


In [275]:
#print the number of rows before merging
rows_before = len(df2)
print("Rows before merge:", rows_before)

Rows before merge: 20973889


In [276]:
#pring the number of rows after merging
print("Rows after merge:", len(df2))

Rows after merge: 20973889


In [277]:
#print the number of columns before merging
columns_before = len(df2.columns)
print("Columns before merge:", columns_before)

Columns before merge: 19


In [278]:
#print the number of rows after merging
print("Columns after merge:", len(df2.columns))

Columns after merge: 19


In [279]:
#Check for duplicate price records
duplicates = df2.duplicated(subset=['ticker', 'date']).sum()

print("Duplicate ticker-date records:", duplicates)

Duplicate ticker-date records: 0


In [280]:
#Verify the merged data
print(
    df2[
        ['ticker', 'date', 'close', 'price_change','daily_return', 'close_lag_1','ma_5', 'ma_20',
         'volatility_20', 'high_low_range','exchange', 'sector', 'industry']
    ].head(20)
)

   ticker       date  close  price_change  daily_return  close_lag_1  ma_5  \
0       A 1999-11-18  31.47           NaN           NaN          NaN   NaN   
1       A 1999-11-19  28.88         -2.59         -8.24        31.47   NaN   
2       A 1999-11-22  31.47          2.59          8.98        28.88   NaN   
3       A 1999-11-23  28.61         -2.86         -9.09        31.47   NaN   
4       A 1999-11-24  29.37          0.76          2.66        28.61 29.96   
5       A 1999-11-26  29.46          0.09          0.30        29.37 29.56   
6       A 1999-11-29  30.13          0.67          2.28        29.46 29.81   
7       A 1999-11-30  30.18          0.04          0.15        30.13 29.55   
8       A 1999-12-01  30.71          0.54          1.78        30.18 29.97   
9       A 1999-12-02  31.56          0.85          2.77        30.71 30.41   
10      A 1999-12-03  31.83          0.27          0.85        31.56 30.88   
11      A 1999-12-06  32.73          0.89          2.81        3

In [282]:
#3.2 Verifying Missing Values Created by Feature Engineering
feature_columns = [
    'price_change',
    'daily_return',
    'close_lag_1',
    'ma_5',
    'ma_20',
    'volatility_20'
]

print(df2[feature_columns].isnull().sum())

price_change       5685
daily_return       5685
close_lag_1        5685
ma_5              22735
ma_20            107884
volatility_20    113555
dtype: int64


I will keep the missing values for now because they represent a lack of historical data, not errors. also if i replace it with 0, it will be missleading

In [283]:
#verify the different type of exhange avavailabe
df2['exchange'].value_counts()

exchange
NYSE      11579051
NASDAQ     9394838
Name: count, dtype: int64

In [284]:
#verify how many unique sectors are there.
df2['sector'].nunique()

12

In [285]:
#verify how many unique industry there are
df2['industry'].nunique()

135

In [286]:
#One-hot encode the categorical variables
df2 = pd.get_dummies(
    df2,
    columns=['exchange', 'sector'],
    drop_first=True,
    dtype=int
)

In [287]:
df2.head()

,ticker,open,close,adj_close,low,high,volume,date,adj_close_flag,price_change,...,sector_CONSUMER DURABLES,sector_CONSUMER NON-DURABLES,sector_CONSUMER SERVICES,sector_ENERGY,sector_FINANCE,sector_HEALTH CARE,sector_MISCELLANEOUS,sector_PUBLIC UTILITIES,sector_TECHNOLOGY,sector_TRANSPORTATION
0,A,32.55,31.47,27.49,28.61,35.77,62546300,1999-11-18,False,NaN,...,0,0,0,0,0,0,0,0,0,0
1,A,30.71,28.88,25.23,28.48,30.76,15234100,1999-11-19,False,-2.59,...,0,0,0,0,0,0,0,0,0,0
2,A,29.55,31.47,27.49,28.66,31.47,6577800,1999-11-22,False,2.59,...,0,0,0,0,0,0,0,0,0,0
3,A,30.40,28.61,25.00,28.61,31.21,5975600,1999-11-23,False,-2.86,...,0,0,0,0,0,0,0,0,0,0
4,A,28.70,29.37,25.66,28.61,30.00,4843200,1999-11-24,False,0.76,...,0,0,0,0,0,0,0,0,0,0


In [288]:
#print 
print(df2.columns)

Index(['ticker', 'open', 'close', 'adj_close', 'low', 'high', 'volume', 'date',
       'adj_close_flag', 'price_change', 'daily_return', 'close_lag_1', 'ma_5',
       'ma_20', 'volatility_20', 'high_low_range', 'industry', 'exchange_NYSE',
       'sector_CAPITAL GOODS', 'sector_CONSUMER DURABLES',
       'sector_CONSUMER NON-DURABLES', 'sector_CONSUMER SERVICES',
       'sector_ENERGY', 'sector_FINANCE', 'sector_HEALTH CARE',
       'sector_MISCELLANEOUS', 'sector_PUBLIC UTILITIES', 'sector_TECHNOLOGY',
       'sector_TRANSPORTATION'],
      dtype='str')


In [291]:
#Final dataset features
final_features = [
    'ticker',
    'open',
    'close',
    'adj_close',
    'low',
    'high',
    'volume',
    'date',
    'price_change',
    'daily_return',
    'close_lag_1',
    'ma_5',
    'ma_20',
    'volatility_20',
    'high_low_range',
    'industry'
]

# Add the one-hot encoded exchange and sector columns
encoded_columns = [
    col for col in df2.columns
    if col.startswith('exchange_') or col.startswith('sector_')
]

final_features = final_features + encoded_columns

final_df = df2[final_features].copy()

print(final_df.head())
print(final_df.shape)

  ticker  open  close  adj_close   low  high    volume       date  \
0      A 32.55  31.47      27.49 28.61 35.77  62546300 1999-11-18   
1      A 30.71  28.88      25.23 28.48 30.76  15234100 1999-11-19   
2      A 29.55  31.47      27.49 28.66 31.47   6577800 1999-11-22   
3      A 30.40  28.61      25.00 28.61 31.21   5975600 1999-11-23   
4      A 28.70  29.37      25.66 28.61 30.00   4843200 1999-11-24   

   price_change  daily_return  ...  sector_CONSUMER DURABLES  \
0           NaN           NaN  ...                         0   
1         -2.59         -8.24  ...                         0   
2          2.59          8.98  ...                         0   
3         -2.86         -9.09  ...                         0   
4          0.76          2.66  ...                         0   

   sector_CONSUMER NON-DURABLES  sector_CONSUMER SERVICES  sector_ENERGY  \
0                             0                         0              0   
1                             0                 

I have excluded adj_close_flag from the final dataset because the values were extremely high and i had initially flagged it. if i include that in the model, the model will place high significance to the high number and therefore provide an unfair decision. Also, if i have to use stanadaization scheme it will take a longer time to complete the standardization

In [293]:
#verify the total columns
print(final_df.columns.tolist())

['ticker', 'open', 'close', 'adj_close', 'low', 'high', 'volume', 'date', 'price_change', 'daily_return', 'close_lag_1', 'ma_5', 'ma_20', 'volatility_20', 'high_low_range', 'industry', 'exchange_NYSE', 'sector_CAPITAL GOODS', 'sector_CONSUMER DURABLES', 'sector_CONSUMER NON-DURABLES', 'sector_CONSUMER SERVICES', 'sector_ENERGY', 'sector_FINANCE', 'sector_HEALTH CARE', 'sector_MISCELLANEOUS', 'sector_PUBLIC UTILITIES', 'sector_TECHNOLOGY', 'sector_TRANSPORTATION']


4.2 Check the Final Dataset, for number of rows/columns, missing values, duplicates, maximum and minimum date


In [294]:
# Number of rows and columns
print("Rows and columns:", final_df.shape)

# Data types
print("\nData types:")
print(final_df.dtypes)

# Missing values
print("\nMissing values:")
print(final_df.isnull().sum())

# Duplicate ticker + date records
print("\nDuplicate ticker + date records:")
print(final_df.duplicated(subset=['ticker', 'date']).sum())

# Minimum and maximum dates
print("\nMinimum date:", final_df['date'].min())
print("Maximum date:", final_df['date'].max())


Rows and columns: (20973889, 28)

Data types:
ticker                                     str
open                                   float64
close                                  float64
adj_close                              float64
low                                    float64
high                                   float64
volume                                   int64
date                            datetime64[us]
price_change                           float64
daily_return                           float64
close_lag_1                            float64
ma_5                                   float64
ma_20                                  float64
volatility_20                          float64
high_low_range                         float64
industry                                   str
exchange_NYSE                            int64
sector_CAPITAL GOODS                     int64
sector_CONSUMER DURABLES                 int64
sector_CONSUMER NON-DURABLES             int64
sector_CONSUME

In [295]:
print(final_df.info())

<class 'pandas.DataFrame'>
RangeIndex: 20973889 entries, 0 to 20973888
Data columns (total 28 columns):
 #   Column                        Dtype         
---  ------                        -----         
 0   ticker                        str           
 1   open                          float64       
 2   close                         float64       
 3   adj_close                     float64       
 4   low                           float64       
 5   high                          float64       
 6   volume                        int64         
 7   date                          datetime64[us]
 8   price_change                  float64       
 9   daily_return                  float64       
 10  close_lag_1                   float64       
 11  ma_5                          float64       
 12  ma_20                         float64       
 13  volatility_20                 float64       
 14  high_low_range                float64       
 15  industry                      str        

There are no duplicates and price_change, daily_return, close_lag_1, ma_5, ma_20, volatility_20, have missing values which is not due to error but last of historical data. Industry however has missing values of 2,549,449. which is realy significat. so i will investigate the values and how much percentage of the entire  dataset is missing and why


In [296]:
print("Total rows:", len(final_df))
print("Missing industry:", final_df['industry'].isna().sum())
print("Percentage missing:", 
      final_df['industry'].isna().mean() * 100)

Total rows: 20973889
Missing industry: 2549449
Percentage missing: 12.155347060337737


12.15% of the total data is missing

In [297]:
#verify which ticker has missing industry
missing_industry = final_df[final_df['industry'].isna()]

print(
    missing_industry['ticker']
    .value_counts()
    .head(20)
)

ticker
ASA     10253
ADX      9696
GAM      9696
TY       9696
JHI      9694
JHS      9694
MCI      9691
WRE      9631
INSI     9613
BIF      9583
MXF      9383
USAU     9331
VBF      8996
KF       8570
ASG      8187
EEA      8091
GAB      8075
ZF       8045
USA      8025
MFM      8008
Name: count, dtype: int64


These are the tickers with missing industry. For now i will not replace them with unknown. unless the machine learning moedel reject NaN

In [298]:
#safe final into csv
final_df.to_csv('stock_prepared.csv', index=False)

In [299]:
final_df.head()

,ticker,open,close,adj_close,low,high,volume,date,price_change,daily_return,...,sector_CONSUMER DURABLES,sector_CONSUMER NON-DURABLES,sector_CONSUMER SERVICES,sector_ENERGY,sector_FINANCE,sector_HEALTH CARE,sector_MISCELLANEOUS,sector_PUBLIC UTILITIES,sector_TECHNOLOGY,sector_TRANSPORTATION
0,A,32.55,31.47,27.49,28.61,35.77,62546300,1999-11-18,NaN,NaN,...,0,0,0,0,0,0,0,0,0,0
1,A,30.71,28.88,25.23,28.48,30.76,15234100,1999-11-19,-2.59,-8.24,...,0,0,0,0,0,0,0,0,0,0
2,A,29.55,31.47,27.49,28.66,31.47,6577800,1999-11-22,2.59,8.98,...,0,0,0,0,0,0,0,0,0,0
3,A,30.40,28.61,25.00,28.61,31.21,5975600,1999-11-23,-2.86,-9.09,...,0,0,0,0,0,0,0,0,0,0
4,A,28.70,29.37,25.66,28.61,30.00,4843200,1999-11-24,0.76,2.66,...,0,0,0,0,0,0,0,0,0,0
